syntax = "proto3";

package raft;

enum RaftState {
    UNKNOWN = 0;
    LEADER = 1;
    FOLLOWER = 2;
    CANDIDATE = 3;
    LEANER = 4;
};

enum RaftLogType {
    LOGTYPE_NORMAL = 0;
    LOGTYPE_ADD_NONVOTING_NODE = 1;
    LOGTYPE_ADD_NODE = 2;
    LOGTYPE_REMOVE_NODE = 3;
    LOGTYPE_PROMOTE_NODE = 4; //promote some node to be leader
};

message LogEntry {
    RaftLogType type = 1;
    uint64 term = 2;
    uint64 index = 3;
    bytes data = 4;
};

message Peer {
    uint64 raftid = 1;
    uint64 nodeid = 2;
    string ip = 3;
    uint32 port = 4;
    RaftState state = 5;
};

/////////////////////////

message VoteRequest {
    uint64 term = 1;
    uint64 candidate = 2; //node id
    uint64 last_term = 3;  //last log term
    uint64 last_index = 4; //last log index
};

message VoteResponse {
    uint64 term = 1;
    uint64 nodeid = 2;
    bool agree = 3; // grant for true or false
};

message AppendEntriesRequest {
    uint64 nodeid = 1;
    uint64 term = 2;
    uint64 commit = 3; // leader commited log index

    uint64 prev_log_term = 4; // last log term
    uint64 prev_log_index = 5; //last log index

    repeated LogEntry entries = 6;
};

message AppendEntriesResponse {
    bool success = 1; //true if match prev_log
    uint64 nodeid = 2;
    uint64 term = 3;
    uint64 current_index = 4;
    uint64 first_index = 5;
};

message MemberChangeRequest {
    RaftLogType type = 1;
    Peer peer = 2;
};

message MemberChangeResponse {
    bool success = 1;
    uint64 term = 2;
    Peer peer = 3;
};

message MembersListRequest {
    uint64 raftid = 1;
};

message MembersListResponse {
    bool success = 1;
    uint64 term = 2;
    uint64 raftid = 3;
    repeated Peer peers = 4;
};

message RaftMessage {
    uint32 raftid = 1;
    oneof msg {
        VoteRequest vt_req  = 2;
        VoteResponse vt_rsp = 3;

        AppendEntriesRequest ae_req  = 4;
        AppendEntriesResponse ae_rsp = 5;

        MemberChangeRequest mc_req  = 6;
        MemberChangeResponse mc_rsp = 7;

        MembersListRequest ml_req  = 8;
        MembersListResponse ml_rsp = 9;
    }
};
